Cleanup & tweak VersionNumber

dwijnand opened the pull request:
builds on #221

dwijnand committed:
Cleanup & tweak VersionNumber

typesafe-tools commented:
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=1.x&zinc=1.x&io=1.x&librarymanagement=pull/222/head&util=1.x&website=1.x&deploy_to_bintray=false&run_sbt_community=false&verbose_dbuild=false&community_verbose=false">this link.</A>

typesafe-tools commented:
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/547/console">is in progress...</a>

typesafe-tools commented:
[//]: # (Refs: 1.x-1.x-1.x-pull/222/head-1.x-1.x)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/547/">has checked</A> the following projects against Scala 2.12,
tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/tree/1.x">1.x                 </A> | sbt/sbt@087f21741d99aa1ae313e06fde7d11d802507abb
  zinc              | <A HREF="https://github.com/sbt/zinc/tree/1.x">1.x                 </A> | sbt/zinc@d6dec01ca95a21fc11f62669b44f1ed4b28ae42d
  io                | <A HREF="https://github.com/sbt/io/tree/1.x">1.x                 </A> | sbt/io@cc3e5332946d6be9e4fd20bb6e6d5b08a04eef8c
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/pull/222">pull/222/head       </A> | sbt/librarymanagement@7c54bd359c4f1ac36241d4eea02d0ab250123042
  util              | <A HREF="https://github.com/sbt/util/tree/1.x">1.x                 </A> | sbt/util@85a30fac4a3f895c5aa6571936bc7d016a0c962c
  website           | <A HREF="https://github.com/sbt/website/tree/1.x">1.x                 </A> | 

:x: The result is: ***FAILED***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=1.x&zinc=1.x&io=1.x&librarymanagement=pull/222/head&util=1.x&website=1.x&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

typesafe-tools commented:
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/548/console">is in progress...</a>

typesafe-tools commented:
[//]: # (Refs: 1.x-1.x-1.x-pull/222/head-1.x-1.x)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/548/">has checked</A> the following projects against Scala 2.12,
tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/tree/1.x">1.x                 </A> | sbt/sbt@087f21741d99aa1ae313e06fde7d11d802507abb
  zinc              | <A HREF="https://github.com/sbt/zinc/tree/1.x">1.x                 </A> | sbt/zinc@d6dec01ca95a21fc11f62669b44f1ed4b28ae42d
  io                | <A HREF="https://github.com/sbt/io/tree/1.x">1.x                 </A> | sbt/io@cc3e5332946d6be9e4fd20bb6e6d5b08a04eef8c
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/pull/222">pull/222/head       </A> | sbt/librarymanagement@7c54bd359c4f1ac36241d4eea02d0ab250123042
  util              | <A HREF="https://github.com/sbt/util/tree/1.x">1.x                 </A> | sbt/util@85a30fac4a3f895c5aa6571936bc7d016a0c962c
  website           | <A HREF="https://github.com/sbt/website/tree/1.x">1.x                 </A> | 

:x: The result is: ***FAILED***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=1.x&zinc=1.x&io=1.x&librarymanagement=pull/222/head&util=1.x&website=1.x&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

dwijnand commented:
@cunei any ideas?

```scala
[error] java.lang.RuntimeException: 
[error] 
[error] Fatal: multiple projects have the same artifacts visible in the same space.
[error] 
[error]   org.scala-sbt#compiler-interface  from sbt212-zinc and sbt212-zinc, both visible in space "sbt212"
[error] 
[error] 	at scala.sys.package$.error(package.scala:27)
[error] 	at com.typesafe.dbuild.model.RepeatableDBuildConfig.<init>(RepeatableBuild.scala:158)
[error] 	at com.typesafe.dbuild.model.RepeatableDBuildConfigH$.fromExtractionOutcome(RepeatableBuild.scala:71)
[error] 	at com.typesafe.dbuild.build.SimpleBuildActor$$anonfun$receive$1$$anonfun$applyOrElse$1$$anonfun$9$$anonfun$apply$12.apply(SimpleBuildActor.scala:171)
[error] 	at com.typesafe.dbuild.build.SimpleBuildActor$$anonfun$receive$1$$anonfun$applyOrElse$1$$anonfun$9$$anonfun$apply$12.apply(SimpleBuildActor.scala:171)
[error] 	at com.typesafe.dbuild.build.SimpleBuildActor$$anonfun$receive$1$$anonfun$applyOrElse$1$$anonfun$9.com$typesafe$dbuild$build$SimpleBuildActor$$anonfun$$anonfun$$anonfun$$nest$1(SimpleBuildActor.scala:160)
An exception occurred.
akka.pattern.AskTimeoutException: Ask timed out on [Actor[akka://dbuild/user/ChainedLoggerSupervisorActor#-1649130644]] after [180000 ms]
	at akka.pattern.PromiseActorRef$$anonfun$1.apply$mcV$sp(AskSupport.scala:334)
	at akka.actor.Scheduler$$anon$7.run(Scheduler.scala:117)
	at scala.concurrent.Future$InternalCallbackExecutor$.scala$concurrent$Future$InternalCallbackExecutor$$unbatchedExecute(Future.scala:694)
	at scala.concurrent.Future$InternalCallbackExecutor$.execute(Future.scala:691)
	at akka.actor.LightArrayRevolverScheduler$TaskHolder.executeTask(Scheduler.scala:474)
	at akka.actor.LightArrayRevolverScheduler$$anon$8.executeBucket$1(Scheduler.scala:425)
	at akka.actor.LightArrayRevolverScheduler$$anon$8.nextTick(Scheduler.scala:429)
	at akka.actor.LightArrayRevolverScheduler$$anon$8.run(Scheduler.scala:381)
	at java.lang.Thread.run(Thread.java:748)
```

cunei commented:
This is due to two subprojects in zinc declaring the same generated artifacts. When dbuild looks at the list, it does not know that only one of the two will publish them, and declares the conflict:
```
sbt:zinc Root> compilerInterfaceJava6Compat/artifacts
[info] * Artifact(compiler-interface, pom, pom, None, Vector(pom), None, Map(), None)
[info] * Artifact(compiler-interface, jar, jar, None, Vector(compile), None, Map(), None)
[info] * Artifact(compiler-interface, src, jar, Some(sources), Vector(compile), None, Map(), None)
[info] * Artifact(compiler-interface, doc, jar, Some(javadoc), Vector(compile), None, Map(), None)
sbt:zinc Root> compilerInterface/artifacts
[info] * Artifact(compiler-interface, pom, pom, None, Vector(pom), None, Map(), None)
[info] * Artifact(compiler-interface, jar, jar, None, Vector(compile), None, Map(), None)
[info] * Artifact(compiler-interface, src, jar, Some(sources), Vector(compile), None, Map(), None)
[info] * Artifact(compiler-interface, doc, jar, Some(javadoc), Vector(compile), None, Map(), None)
```
For zinc in particular, one way out is checking the value of `skip in publish` within dbuild; that flag is only available in sbt since 1.0.0-RC3, and dbuild has not been updated to take it into account, yet. Tracking that as: https://github.com/lightbend/dbuild/issues/205

cunei commented:
The `skip in publish` issue is fixed in https://github.com/lightbend/dbuild/issues/206, I updated the validator with a new dbuild version. @dwijnand

typesafe-tools commented:
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/555/console">is in progress...</a>

typesafe-tools commented:
[//]: # (Refs: 1.x-1.x-1.x-pull/222/head-1.x-1.x)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/555/">has checked</A> the following projects against Scala 2.12,
tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/tree/1.x">1.x                 </A> | 
  zinc              | <A HREF="https://github.com/sbt/zinc/tree/1.x">1.x                 </A> | 
  io                | <A HREF="https://github.com/sbt/io/tree/1.x">1.x                 </A> | 
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/pull/222">pull/222/head       </A> | 
  util              | <A HREF="https://github.com/sbt/util/tree/1.x">1.x                 </A> | 
  website           | <A HREF="https://github.com/sbt/website/tree/1.x">1.x                 </A> | 

:x: The result is: ***FAILED***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=1.x&zinc=1.x&io=1.x&librarymanagement=pull/222/head&util=1.x&website=1.x&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

dwijnand commented:
thank you Toni! how do we get jenkins to fetch the new version?

```
Error: Could not retrieve build 0.9.12-toni3
```

typesafe-tools commented:
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/556/console">is in progress...</a>

typesafe-tools commented:
[//]: # (Refs: 1.x-1.x-1.x-pull/222/head-1.x-1.x)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/556/">has checked</A> the following projects against Scala 2.12,
tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/tree/1.x">1.x                 </A> | sbt/sbt@933668076a2ce76680490273fc7dfadcf18d7dbd
  zinc              | <A HREF="https://github.com/sbt/zinc/tree/1.x">1.x                 </A> | sbt/zinc@d6dec01ca95a21fc11f62669b44f1ed4b28ae42d
  io                | <A HREF="https://github.com/sbt/io/tree/1.x">1.x                 </A> | sbt/io@cc3e5332946d6be9e4fd20bb6e6d5b08a04eef8c
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/pull/222">pull/222/head       </A> | sbt/librarymanagement@7c54bd359c4f1ac36241d4eea02d0ab250123042
  util              | <A HREF="https://github.com/sbt/util/tree/1.x">1.x                 </A> | sbt/util@85a30fac4a3f895c5aa6571936bc7d016a0c962c
  website           | <A HREF="https://github.com/sbt/website/tree/1.x">1.x                 </A> | 

:white_check_mark: The result is: ***SUCCESS***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=1.x&zinc=1.x&io=1.x&librarymanagement=pull/222/head&util=1.x&website=1.x&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

dwijnand commented:
nice, validation passed. thanks Toni for unblocking that.

cunei commented:
Sure no problems. I just tweaked the config, should be ok now.

dwijnand commented:
rebased on the final merged form of VersionNumberSpec in #221

eed3si9n commented on core/src/test/scala/sbt/librarymanagement/VersionNumberSpec.scala:
```diff
@@ -51,7 +51,7 @@ class VersionNumberSpec extends FreeSpec with Matchers with Inside {
     assertIsNotCompatibleWith(v, "0.12.0-RC1", SecondSegment)
     assertIsCompatibleWith(v, "0.12.1", SecondSegment)
-    assertIsCompatibleWith(v, "0.12.1-M1", SecondSegment)
+    assertIsNotCompatibleWith(v, "0.12.1-M1", SecondSegment)
```
Why did this change?

eed3si9n commented on core/src/test/scala/sbt/librarymanagement/VersionNumberSpec.scala:
```diff
@@ -85,7 +85,7 @@ class VersionNumberSpec extends FreeSpec with Matchers with Inside {
     assertParsesTo(v, Seq(2, 10, 4), Seq("20140115", "000117", "b3a", "sources"), Seq())
     assertCascadesTo(v, Seq("2.10.4-20140115-000117-b3a-sources", "2.10.4", "2.10"))
-    assertIsCompatibleWith(v, "2.0.0", SemVer)
+    assertIsNotCompatibleWith(v, "2.0.0", SemVer)
```
According to SemVer 2.10.4-20140115-000117-b3a-sources should be compat with 2.0.0, right?

eed3si9n commented on core/src/main/scala/sbt/librarymanagement/VersionNumber.scala:
```diff
@@ -5,159 +5,188 @@ final class VersionNumber private[sbt] (
+      if (v.extras.isEmpty) v else VersionNumber(v.numbers, v.tags, Nil)
+
+    // SemVer Spec Rule 9 (above)
+    private[VersionNumber] def isPrerelease(v: VersionNumber): Boolean = v.tags.nonEmpty
```
This doesn't capture the subtlety of Rule 9.

> Pre-release versions have a lower precedence than the associated normal version.

Here are some examples:

- 1.0.0-M1 is a pre-release of 1.0.0, and thus have lower precedence than 1.0.0.
- 1.1.0-M1 is a pre-release of 1.1.0, and thus have lower precedence than 1.1.0. However, since this is intended to be a pre-release of 1.1.0 series, a reasonable expectation is to assume compatibility with 1.0.0 series.

Thus my original code treats `case (_, 0, 0, ts1, _, 0, 0, ts2) if ts1.nonEmpty || ts2.nonEmpty` as a special case.

dwijnand replied:
> However, since this is intended to be a pre-release of 1.1.0 series, a reasonable expectation is to assume compatibility with 1.0.0 series.

SemVer states:

> A pre-release version indicates that the version is unstable and might not satisfy the intended compatibility requirements as denoted by its associated normal version.

eed3si9n replied:
Right, so in case of 1.0.0-M1 whose associated normal version is 1.0.0, the "might not" actually is meaningful. But this doesn't mean that every time we issue an M1, it's completely incompatible with sbt 1.x for example.

dwijnand replied:
I'll push a commit with the diff of allowing these to be compatible, so we can compare.

dwijnand committed:
Allow pre-release versions to be compatible with normal versions

dwijnand committed:
Change semantics to match previous semantics

We're unsure about SemVer's semantics around pre-releases, so for now we
just revert this refactoring to preserve the semantics it's always had.
